man/dnscache.8
.TH djbdnscurve6: dnscache 8
.SH NAME
dnscache \- DNS cache server and iterative resolver
.SH DESCRIPTON
.B dnscache
accepts recursive DNS queries
from local clients such as web browsers and mail transfer agents.
It collects responses from remote DNS servers.
It caches the responses for faster client lookup.

Upon sending iterative queries
.B dnscache
will use the received FQDN of authoritive name servers to
evalute their potential usage for
encryption in case the hostname starts with
.IR uz5 .
Subsequently 
.B dnscache
will automatically encrypt queries to those servers 
identified to be DNSCurve capable.
Now, both queries and responses are encrypted 
using either the propriatory
.I DNSCurve stream
format, or if the environment variable
.I $USETXTFORMAT
is set, the less performant 
.I TXT 
format to cope with deep packet 
inspecting Firewalls analysing DNS 
traffic and expecting a standard DNS header.

In case the name server does not
respond to encrypted UDP queries, 
.B dnscache 
falls back to unencrypted queries.

.B dnscache
may be used in a
.I forward only 
mode. 
.SH "CONFIGURATION"
Normally 
.B dnscache
is set up by the
.B dnscache-conf
program.

.B dnscache
runs chrooted in the directory
specified by the 
.I $ROOT
environment variable,
under the uid and gid
specified by the 
.I $UID
and 
.I $GID
environment variables.
In case 
.I $USETEXTFORMAT
is set, 
.B dnscache
will send queries even for DNSCurve
enabled name servers using standard
.I DNS TXT 
headers.

.B dnscache
can be adviced to go to unencrypted fallback mode, 
if hostname of the name server starts with 
.I uz5
but don't respond to encrypted UDP queries
for this lookup.
This behavior can be changed using the 
environment variable
.IR UZ5FALLBACK=n .
A value like
.I n=1
might impact correctly behaving DNSCurve
name servers which do not respond to the
initial query, while larger values like
.I n=3
delays name resolution for those name
servers significantly. 
Setting
.I UZ5FALLBACK=0
disables fallback mode, which is the default.

The value
.I UZ5FALLBACK=2
is initally set-up by
.BR dnscache-conf .

.SH "BINDINGS"
.B dnscache
listens for incoming UDP packets and TCP connections
addressed to port 53 for
.IR $IP ,
which could be either an IPv4 or IPv6 scoped addresss, 
supporting both public or internal access.

Given a host scope, one typically uses the addresses
.IR 127.0.0.1 ,
.IR ::1 , 
or 
.IR fe80::1%lo0 .
In those cases,
.B dnscache
serves the own host only.

Setting up 
.B dnscache
on a private network requires private IPv4 addresses;
while for IPv6 ULA and LLU addresses can be used. Examples:
.IR 10.10.10.53 ,
.IR fd00::53 ,
.IR fe80::53%eth0 .

.B dnscache 
is able to serve all existing IP addresses on the host (multi-homing).
For IPv4 specify
.I 0.0.0.0
and for IPv6 set 
.I ::
within
.IR env/IP .

In case of 
.I :: 
.B dnscache 
additionally supports reverse anycasting for IPv6.  Now,
.B dnscache 
will accept IPv6 packets from every available interface,
even if dynamically allocated.

.B dnscache 
forces simultaneous bind to IPv4 and IPv6 addresses
in case a 'pseudo' IP address is specified as
.IR :0 .
However, this will not trigger reverse anycasting support.

.B dnscache
sends outgoing packets from high ports of 
.IR $IPSEND .
Typically 
.I $IPSEND
is 
.I 0.0.0.0 
or
.I ::
meaning the machine's primary IP address
covering both IPv4 and IPv6.
However, a specific sending IP address can be used,
which might be destinct from the receiving ones. 
.SH "FOWARDING MODE"
If
.I $FORWARDONLY
is set,
.B dnscache
treats
.I servers/@
as a list of IP addresses
for other caches, not root servers.
It forwards queries to those caches.
It does not contact the root servers, 
or any other DNS servers, directly.
.SH "CLIENT QUALIFICATION"
.B dnscache
accepts a packet or connection
from the IPv4 address
.I 1.2.3.4
if it sees a file named
.I ip/1.2.3.4
or
.I ip/1.2.3
or
.I ip/1.2
or
.IR ip/1 .
For IPv6 addresses
.B dnscache
can be instructed in a similar way:
.IR ip/2001::fefe ,
.IR ip/2001:a:b:c:d ,
.IR ip/2001 ,
.I ip/fe80 
(all LLU),
.I ip/fd00 
(all ULA).

.B dnscache
will reject packets or connections
from IP addresses marked as 'commented out':
.IR ip/#2001::fec ,
.IR ip/#192.168.1 .
Rejections have precedence over acceptance.
 Note: In any case, the delimiter
(either '.' or ':') shall not be used as last character.
.SH "SERVER QUALIFICATION"
If
.B dnscache
recognizes the environment variable
.IR $FLAGEDSERVER ,
name server listed under
.I ip/
are treated in the following way:
Servers included as
.I ip/%1.2.3.4 
or 
.I ip/%2001::a:b:c:d
and given their dotted-decimal IPv4
or compactified IPv6 addresses are omitted 
for name resolution in case the IP address
is prepended with a
.IR % .
If the IP addresses is prepended with a
.IR - ,
rather instead of a DNSCurve query a standard
query will be used, irrespectively if the 
server's FQDN starts with the magic 
.IR uz5 .
Example:

In case
.I $IP4 
is set,
.B dnscache
will contact only nameserves given their IPv4 address.

.SH "ROOT SERVERS"
.B dnscache
reads a seed, up to 128 bytes,
from standard input,
and passes the seed to
.BR dns_random_init .

.B dnscache
reads a list of root server given as dotted-decimal IPv4 
and/or compactified IPv6 addresses one per line 
from the file
.IR servers/@ .

A total of 32 names servers is handled, which are
specified in dotted-decimal IPv4 or compactified IPv6 format.  
Name severs specified by their IPv6 LLU addresses need to 
include the interface name via those they are reachable.

.B dnscache
also scans the 
.I servers
directory
for server IP addresses for other domains.
If there are addresses listed in 
.IR servers/moon.af.mil ,
for example,
then 
.B dnscache
will send queries for 
.I anything.moon.af.mil
to those addresses,
and will not cache records for 
.I anything.moon.af.mil
from outside servers such as the root servers.
.SH "MEMORY USE"
.B dnscache
uses a fixed-size table, under 512K,
to keep track of as many as 400 simultaneous UDP queries
and 40 simultaneous TCP connections.
It also dynamically allocates memory,
usually just a few bytes but occasionally much more,
for each active query.
If it runs out of memory handling a query, it discards that query.

.B dnscache
asks the operating system to reserve a 128K buffer
for bursts of incoming UDP queries.
If a new UDP query arrives
when
.B dnscache
is already handling 400 simultaneous UDP queries,
.B dnscache
drops the oldest query.
If a new TCP connection arrives
when
.B dnscache
is already handling 40 simultaneous TCP connections,
.B dnscache
drops the oldest connection.

.B dnscache
uses a fixed-size cache,
as controlled by the 
.I $CACHESIZE
environment variable.
Roughly 5% of the cache is used for a hash table.
The rest is used for cache entries
(including 8-byte Y2038-compliant expiration times):

.TP
o
A sets.
22 bytes plus 4 bytes per address plus the length of the owner name.
.TP
o
AAAA sets.
22 bytes plus 16 bytes per address plus the length of the owner name.
.TP
o
NS sets or PTR sets or CNAME sets.
22 bytes plus the length of the owner name and all the data names.
.TP
o
MX sets.
22 bytes plus 2 bytes per MX plus the length of all the names.
.TP
o
Other record sets.
22 bytes plus 2 bytes per record
plus the length of all the data strings
plus the length of the owner name.
.TP
o
Nonexistent domain or server failure.
22 bytes plus the length of the owner name.

.P
Sets larger than 8192 bytes are not cached.

.B dnscache
does not exit when it runs out of space in its cache;
it simply removes the oldest entries to make more space.
.SH "MESSAGE SIZES"
.B dnscache 
is expecting to be used on IPv6 capabable networks
supporting a 'minimum length' MLMTU size of 1280 byte (RFC 8200)
allowing larger UDP packet sizes than for IPv4 only.
Upon start,
.B dnscache 
shows the UDP message size supported by default.
In addition,
.B dnscache
understands EDNS(0) extensions in DNS messages (RFC 6891), 
typically used by DNSSEC.
.SH "RESOLUTION AND CACHING POLICIES"
.B dnscache
relies on a configured list of root name servers.
However, the IP addresses of the Internet root servers are subject of change.

.B dnscache
does not cache (or pass along) records outside the server's bailiwick;
those records could be poisoned.
Records for 
.IR foo.dom ,
for example,
are accepted only from the root servers,
the 
.I dom
servers, and the 
.I foo.dom
servers.

.B dnscache
does not bypass its cache
to obtain glue from the additional section of a response.
In particular, it will not use glue outside the server's bailiwick,
or glue with TTL 0,
or glue that violates other caching policies.

.B dnscache
caches records for at most a week.
It interprets TTLs above 2147483647 as 0.

.B dnscache
does not cache SOA records.
However, it does use SOA TTLs to determine cache times (up to an hour)
for zero-record responses and nonexistent domains.
.SH "RESPONSES TO QUERIES"
.BR dnscache 's
responses are generally much smaller than BIND's responses.
They do not include authority records
(NS records of the source name servers
and SOA records for negative answers)
or additional records
(A records relevant to NS or MX records).
When the answer section is truncated by UDP length limits,
it is eliminated entirely.

.B dnscache
tries to prevent local users from snooping on other local users.
It discards non-recursive queries;
it discards inverse queries;
and it discards zone-transfer requests.
If
.I $HIDETTL
is set,
.B dnscache
always uses a TTL of 0 in its responses.

According to RFC 1035,
the AA bit ``specifies that the responding name server 
is an authority for the domain name in question section.''

.B dnscache
is not an authority for any domain names.

.B dnscache
never sets the AA bit
(except in NXDOMAIN responses, as required by RFC 2308,
to work around a common client bug).
In contrast, BIND often sets AA for positive responses
even when it is not an authority for the domain name.
.SH "SPECIAL NAMES"
.B dnscache
handles 
.I localhost
internally,
giving it an A record of 127.0.0.1.
In addition, for the IPv6 address ::1
it considers those as
.I ipv6-localhost 
together with the respective AAAA record.
.B dnscache
handles 
.I 1.0.0.127.in-addr.arpa
and
.I 1.0.0...0.ip6.arpa
internally,
giving it a PTR record of 127.0.0.1 and ::1 respectively.

.B dnscache
handles dotted-decimal domain names internally,
giving (e.g.) the domain name 
.I 192.48.96.2
an A record of 
.IR 192.48.96.2 .
.SH "SEE ALSO"
dnscache-conf(8),
dnscache-log(5)
.SH REFERENCE
http://cr.yp.to/djbdns.html


src/dnscache.c
#include <unistd.h>
#include "case.h"
#include "env.h"
#include "exit.h"
#include "scan.h"
#include "logmsg.h"
#include "str.h"
#include "ip.h"
#include "uint_t.h"
#include "socket_if.h"
#include "dns.h"
#include "taia.h"
#include "byte.h"
#include "roots.h"
#include "fmt.h"
#include "iopause.h"
#include "query.h"
#include "alloc.h"
#include "response.h"
#include "sipcache.h"
#include "ndelay.h"
#include "log.h"
#include "clientok.h"
#include "droproot.h"
#include "readclose.h"
#include "curvedns.h"

#define WHO "dnscache"

uint16 dnsport = 53;
stralloc ifname = {0};
uint32 ifidx = 0; /* aka scope_id */

int flagusetxtformat = 0;
int flagdualstack = 0;
int flagipv6anycast = 0;
char *p = "u";

static int packetquery(char *buf,unsigned int len,char **q,char qtype[2],char qclass[2],char id[2])
{
  unsigned int pos;
  unsigned int ad = 0;
  char header[12];

  errno = EPROTO;
  pos = dns_packet_copy(buf,len,0,header,12); if (!pos) return 0;
  if (header[2] & 128) return 0;  /* must not respond to responses */
  if (!(header[2] & 1)) return 0; /* do not respond to non-recursive queries */
  if (header[2] & 120) return 0;  /* standard query only */
  if (header[2] & 2) return 0;    /* Truncation not allowed */
  if (header[3] & 32) ad = 1; 	  /* Authenticated data */
  if (byte_diff(header + 4,2,"\0\1")) return 0;

  pos = dns_packet_getname(buf,len,pos,q); if (!pos) return 0;
  pos = dns_packet_copy(buf,len,pos,qtype,2); if (!pos) return 0;
  pos = dns_packet_copy(buf,len,pos,qclass,2); if (!pos) return 0;
  if (byte_diff(qclass,2,DNS_C_IN) && byte_diff(qclass,2,DNS_C_ANY)) return 0;
  if (!dns_packet_edns0(header,buf,len,pos)) return 0;

  byte_copy(id,2,header);
  return 1;
}

static char ipsending[16];
static char iplistening[16];
// static char buf[2 * MSGSIZE + 2];
static char buf[MAXSEGMENT];
uint64 numqueries = 0;

static int udp53;

static struct udpclient {
  struct query q;
  struct taia start;
  uint64 active; /* query number, if active; otherwise 0 */
  iopause_fd *io;
  char ip[16];
  uint16 port;
  char id[2];
  uint32 scope_id;
} u[QUERY_MAXUDP];

int uactive = 0;
int eactive = 0;

void u_drop(int j)
{
  if (!u[j].active) return;
  p = "u";
  log_querydrop(&u[j].active,p);
  u[j].active = 0; --uactive; 
}

void u_respond(int j)
{
  if (!u[j].active) return;
  response_id(u[j].id);
  if (response_len > MSGSIZE) response_tc();
    socket_send(udp53,response,response_len,u[j].ip,u[j].port,u[j].scope_id);
  p = "u";
  log_querydone(&u[j].active,response_len,p);
  u[j].active = 0; --uactive; 
}

void u_new(void)
{
  int j;
  int i;
  struct udpclient *x;
  int len;
  static char *q = 0;
  char qtype[2];
  char qclass[2];

  for (j = 0; j < QUERY_MAXUDP; ++j)
    if (!u[j].active)
      break;

  if (j >= QUERY_MAXUDP) {
    j = 0;
    for (i = 1; i < QUERY_MAXUDP; ++i)
      if (taia_less(&u[i].start,&u[j].start))
        j = i;
        errno = ETIMEDOUT;
        u_drop(j);
  }  

  x = u + j;
  taia_now(&x->start);

  len = socket_recv(udp53,buf,sizeof(buf),x->ip,&x->port,&x->scope_id);
  if (len == -1) return;
  if (len >= sizeof(buf)) return;
  if (x->port < 1024) if (x->port != 53) return;
  if (!clientok(x->ip)) { 
    char ipstr[IP6_FMT];
    if (ip6_isv4mapped(x->ip)) len = ip4_fmt(ipstr,x->ip + 12);
    else len = ip6_fmt(ipstr,x->ip);
    ipstr[len] = '\0';
    logmsg(WHO,-99,WARN,B("client blocked: ",ipstr)); 
    return;
  }

  if (!packetquery(buf,len,&q,qtype,qclass,x->id)) return;

  x->active = ++numqueries; ++uactive;
  p = "u";
  if (len > MINMSGSIZE) { ++eactive; p = "e"; }
  log_query(&x->active,x->ip,x->port,x->id,q,qtype,p);

  switch (query_start(&x->q,q,qtype,qclass,ipsending,&x->scope_id)) {
    case -1: case -2: u_drop(j); return;
    case 1: u_respond(j);
  }
}

static int tcp53;

struct tcpclient {
  struct query q;
  struct taia start;
  struct taia timeout;
  uint64 active; /* query number or 1, if active; otherwise 0 */
  iopause_fd *io;
  char ip[16]; /* send response to this address */
  uint16 port; /* send response to this port */
  char id[2];
  int tcp; /* open TCP socket, if active */
  int state;
  char *buf; /* 0, or dynamically allocated of length len */
  unsigned int len;
  unsigned int pos;
  uint32 scope_id;
} t[QUERY_MAXTCP];
int tactive = 0;

/*
state 1: buf 0; normal state at beginning of TCP connection
state 2: buf 0; have read 1 byte of query packet length into len
state 3: buf allocated; have read pos bytes of buf
state 0: buf 0; handling query in q
state -1: buf allocated; have written pos bytes
*/

void t_free(int j)
{
  if (!t[j].buf) return;
  alloc_free(t[j].buf);
  t[j].buf = 0;
}

void t_timeout(int j)
{
  struct taia now;
  if (!t[j].active) return;
  taia_now(&now);
  taia_uint(&t[j].timeout,10);
  taia_add(&t[j].timeout,&t[j].timeout,&now);
}

void t_close(int j)
{
  if (!t[j].active) return;
  t_free(j);
  log_tcpclose(t[j].ip,t[j].port);
  close(t[j].tcp);
  t[j].active = 0; --tactive;
}

void t_drop(int j)
{
  p = "t";
  log_querydrop(&t[j].active,p);
  errno = EPIPE;
  t_close(j);
}

void t_respond(int j)
{
  if (!t[j].active) return;
  p = "t";
  log_querydone(&t[j].active,response_len,p);
  response_id(t[j].id);
  t[j].len = response_len + 2;
  t_free(j);
  t[j].buf = alloc(response_len + 2);
  if (!t[j].buf) { t_close(j); return; }
  uint16_pack_big(t[j].buf,response_len);
  byte_copy(t[j].buf + 2,response_len,response);
  t[j].pos = 0;
  t[j].state = -1;
}

void t_rw(int j)
{
  struct tcpclient *x;
  char *ch;
  static char *q = 0;
  unsigned int readsize;
  char qtype[2];
  char qclass[2];
  int r;

  x = t + j;
  if (x->state == -1) {
    r = write(x->tcp,x->buf + x->pos,x->len - x->pos);
    if (r <= 0) { t_close(j); return; }
    x->pos += r;
    if (x->pos == x->len) {
      t_free(j);
      x->state = 1; /* could drop connection immediately */
    }
    return;
  }

  switch (x->state) {
    case 1: readsize = 2U; break;
    case 2: readsize = 1U; break;
    case 3: readsize = x->len - x->pos; break;
    default: return; /* impossible */
  }

  r = read(x->tcp,buf,readsize);
  if (r == 0) { errno = EPIPE; t_close(j); return; }
  if (r < 0) { t_close(j); return; }

  ch = buf;
  if (x->state == 1) {
    x->len = (unsigned char) *ch++;
    x->len <<= 8;
    x->state = 2;
    if (--r <= 0) return;
  }
  if (x->state == 2) {
    x->len += (unsigned char) *ch;
    if (!x->len) { errno = EPIPE; t_close(j); return; }
    x->buf = alloc(x->len);
    if (!x->buf) { t_close(j); return; }
    x->pos = 0;
    x->state = 3;
    return;
  }

  if (x->state != 3) return; /* impossible */

  byte_copy(&x->buf[x->pos],r,ch);
	x->pos += r;
  if (x->pos < x->len) return;

  if (!packetquery(x->buf,x->len,&q,qtype,qclass,x->id)) { t_close(j); return; }

  x->active = ++numqueries;
  p = "t";
  log_query(&x->active,x->ip,x->port,x->id,q,qtype,p);
  switch (query_start(&x->q,q,qtype,qclass,ipsending,ifidx)) {
    case -1: case -2: t_drop(j); return;
    case 1: t_respond(j); return;
  }
  t_free(j);
  x->state = 0;
}

void t_new(void)
{
  int i;
  int j;
  int len;
  struct tcpclient *x;

  for (j = 0; j < QUERY_MAXTCP; ++j)
    if (!t[j].active)
      break;

  if (j >= QUERY_MAXTCP) {
    j = 0;
    for (i = 1; i < QUERY_MAXTCP; ++i)
      if (taia_less(&t[i].start,&t[j].start))
        j = i;
    errno = ETIMEDOUT;
    if (t[j].state == 0)
      t_drop(j);
    else
      t_close(j);
  }

  x = t + j;
  taia_now(&x->start);

  x->tcp = socket_accept(tcp53,x->ip,&x->port,&x->scope_id);
  if (x->tcp == -1) return;
  if (x->port < 1024) if (x->port != 53) { close(x->tcp); return; }
  if (!clientok(x->ip)) { 
    char ipstr[IP6_FMT];
    if (ip6_isv4mapped(x->ip)) len = ip4_fmt(ipstr,x->ip + 12);
    else len = ip6_fmt(ipstr,x->ip);
    ipstr[len + 1] = '\0';
    logmsg(WHO,-99,WARN,B("client blocked: ",ipstr));
    close(x->tcp); 
    return; 
  }
  if (ndelay_on(x->tcp) == -1) { close(x->tcp); return; } /* Linux bug */

  x->active = 1; ++tactive;
  x->state = 1;
  t_timeout(j);

  log_tcpopen(x->ip,x->port);
}


iopause_fd io[3 + QUERY_MAXUDP + QUERY_MAXTCP];
iopause_fd *udp53io;
iopause_fd *tcp53io;

static void doit(void)
{
  int j;
  struct taia deadline;
  struct taia stamp;
  int iolen;
  int r;

  for (;;) {
    taia_now(&stamp);
    taia_uint(&deadline,120);
    taia_add(&deadline,&deadline,&stamp);

    iolen = 0;

    udp53io = io + iolen++;
    udp53io->fd = udp53;
    udp53io->events = IOPAUSE_READ;

    tcp53io = io + iolen++;
    tcp53io->fd = tcp53;
    tcp53io->events = IOPAUSE_READ;

    for (j = 0; j < QUERY_MAXUDP; ++j)
      if (u[j].active) {
        u[j].io = io + iolen++;
        query_io(&u[j].q,u[j].io,&deadline);
      }

    for (j = 0; j < QUERY_MAXTCP; ++j)
      if (t[j].active) {
        t[j].io = io + iolen++;
        if (t[j].state == 0) 
           query_io(&t[j].q,t[j].io,&deadline);
        else {
           if (taia_less(&t[j].timeout,&deadline)) deadline = t[j].timeout;
           t[j].io->fd = t[j].tcp;
           t[j].io->events = (t[j].state > 0) ? IOPAUSE_READ : IOPAUSE_WRITE;
        }
      }

    if (iopause(io,iolen,&deadline,&stamp) < 0) {
        errno = ECONNRESET;
        logmsg(WHO,errno,FATAL,"IO resources not available");
    }

    for (j = 0; j < QUERY_MAXUDP; ++j)
      if (u[j].active) {
        r = query_get(&u[j].q,u[j].io,&stamp);
        if (r == -1 || r == -2 || r == -3) { errno = ECONNRESET; u_drop(j); }
        if (r == 1) u_respond(j);
      }

    for (j = 0; j < QUERY_MAXTCP; ++j)
      if (t[j].active) {
        if (t[j].io->revents) t_timeout(j);
        if (t[j].state == 0) {
          r = query_get(&t[j].q,t[j].io,&stamp);
          if (r == -1 || r == -2 || r == -3) { errno = ECONNRESET; t_drop(j); }
          if (r == 1) t_respond(j);
        }
        else if (t[j].io->revents || taia_less(&t[j].timeout,&stamp))
          t_rw(j);
      }

    if (udp53io)
      if (udp53io->revents)
        u_new();

    if (tcp53io)
      if (tcp53io->revents)
        t_new();
  }
}
  
char seed[128];

int main()
{
  char *x;
  unsigned long cachesize;
  flagedserver = 0;
  fallback = 0;

  x = env_get("IP");
  if (!x)
    logmsg(WHO,111,ERROR,"$IP not set"); 
  if (case_equals(x,"::")) {
    flagipv6anycast = 1;
  } else if (case_equals(x,":0")) {
    flagdualstack = 1;
    byte_copy(x,2,"::");
    ifidx = 0;
  }
  if (!ip6_ifscan(x,iplistening,&ifname))
    logmsg(WHO,101,SYNTAX,B("unable to parse IP address: ",x));

  if (ifname.len > 1) ifidx = socket_getifidx(ifname.s);

  if (ip6_isv4mapped(iplistening))
    udp53 = socket_udp4();
  else 
    udp53 = socket_udp();
  if (udp53 == -1)
    logmsg(WHO,111,FATAL,"unable to create UDP socket");

  if (flagdualstack) socket_dualstack(udp53);
  if (flagipv6anycast) socket_ip6anycast(udp53);
  if (socket_bind_reuse(udp53,iplistening,dnsport,ifidx) == -1)
    logmsg(WHO,111,FATAL,"unable to bind to UDP socket");

  if (ip6_isv4mapped(iplistening))
    tcp53 = socket_tcp4();
  else 
    tcp53 = socket_tcp6();
  if (tcp53 == -1)
    logmsg(WHO,111,FATAL,"unable to create TCP socket");

  if (flagdualstack) socket_dualstack(tcp53);
  if (flagipv6anycast) socket_ip6anycast(tcp53);
  if (socket_bind_reuse(tcp53,iplistening,dnsport,ifidx) == -1)
    logmsg(WHO,111,FATAL,"unable to bind to TCP socket");

  droproot(WHO);

  socket_tryreservein(udp53,131072);

  byte_zero(seed,sizeof(seed));
  read(0,seed,sizeof(seed));
  dns_random_init(seed);
  close(0);
  query_init();

  x = env_get("IPSEND");
  if (!x)
    logmsg(WHO,111,ERROR,"$IPSEND not set"); 
  if (!ip6_ifscan(x,ipsending,&ifname))
    logmsg(WHO,100,SYNTAX,B("unable to parse IP address: ",x));

  x = env_get("CACHESIZE");
  if (!x)
    logmsg(WHO,111,ERROR,"$CACHESIZE not set"); 
  scan_ulong(x,&cachesize);
  if (!cache_init(cachesize))
    logmsg(WHO,111,FATAL,B("not enough memory for cache of size: ",x));

  if (env_get("HIDETTL"))
    response_hidettl();
  if (env_get("FORWARDONLY"))
    query_forwardonly();
  if (env_get("USETXTFORMAT"))
    flagusetxtformat = 1;
  if (env_get("USETEXTFORMAT"))
    flagusetxtformat = 1;
  x =  env_get("UZ5FALLBACK");
  if (x) scan_uint(x,&fallback);
  if (env_get("FLAGEDSERVER"))
    flagedserver = 1;


  if (!roots_init())
    logmsg(WHO,111,ERROR,"unable to read servers");

  if (socket_listen(tcp53,TCP_BACKLOG) == -1)
    logmsg(WHO,111,FATAL,"unable to listen on TCP socket");

  log_startup(iplistening,ifidx,ipsending,MSGSIZE);
  doit();
}
